pkg/controller/persistentvolume/persistentvolume_claim_binder_controller.go

```diff
@@ -282,20 +283,40 @@ func syncClaim(volumeIndex *persistentVolumeOrderedIndex, binderClient binderCli
 			return fmt.Errorf("A volume match does not exist for persistent claim: %s", claim.Name)
 		}
 
-		// make a binding reference to the claim.
-		// triggers update of the claim in this controller, which builds claim status
+		// create a reference to the claim and assign it to the volume being bound.
+		// the volume is a pointer and assigning the reference fixes a race condition where another
+		// claim might match this volume but before the claimRef is persistent in the next case statement
+		claimRef, err := api.GetReference(claim)
+		if err != nil {
+			return fmt.Errorf("Unexpected error getting claim reference: %v\n", err)
+		}
+
+		// make a binding reference to the claim and ensure to update the local index to prevent dupe bindings
+		clone, err := conversion.NewCloner().DeepCopy(volume)
+		if err != nil {
+			return fmt.Errorf("Error cloning pv: %v", err)
+		}
+		volumeClone, ok := clone.(*api.PersistentVolume)
+		if !ok {
+			return fmt.Errorf("Unexpected pv cast error : %v\n", volumeClone)
+		}
+		volumeClone.Spec.ClaimRef = claimRef
+		if updatedVolume, err := binderClient.UpdatePersistentVolume(volumeClone); err != nil {
+			return fmt.Errorf("Unexpected error saving PersistentVolume.Status: %+v", err)
+		} else {
+			volumeIndex.Update(updatedVolume)
+		}
+
+		// the bind is persisted on the volume above and will always match the claim in a search.
+		// claim would remain Pending if the update fails, but processing this state is idempotent.
 		claim.Spec.VolumeName = volume.Name
-		// TODO: make this similar to Pod's binding both with BindingREST subresource and GuaranteedUpdate helper in etcd.go
 		claim, err = binderClient.UpdatePersistentVolumeClaim(claim)
```

Comment:
> You're updating status in this flow anyway. I don't see reason to avoid updating all the values atomically in this branch. If your status update fails, you'll end up back here anyway.

The line changed, but I think that this
```go
		if len(claim.Status.AccessModes) == 0 {
			claim.Status.Phase = api.ClaimBound
			claim.Status.AccessModes = volume.Spec.AccessModes
			claim.Status.Capacity = volume.Spec.Capacity
			_, err := binderClient.UpdatePersistentVolumeClaimStatus(claim)
			if err != nil {
				return fmt.Errorf("Unexpected error saving claim status: %+v", err)
			}
		}
```

Should be done here, unconditionally stomp what's set, set the phase, and do the update.  I think that doing it here, obviates the need for `ClaimBound`